File: Periodo3/aeds2/estudo/atvLab/Main.java

import java.util.*;

class Arvore {

    No raiz;

    public Arvore() {
        raiz = null;
    }

    public void inserir(int x) {
        if (raiz == null) {
            raiz = new No(x);
        } else {
            inserir(x, raiz);
        }
    }

    private void inserir(int x, No no) {
        if (noDiferente(no.elemento, x)) {
            if (no.elemento > x) {
                // elemento menor vai pra esquerda
                if (no.esq == null) {
                    no.esq = new No(x);
                } else {
                    inserir(x, no.esq);
                }
            } else {
                if (no.dir == null) {
                    no.dir = new No(x);
                } else {
                    inserir(x, no.dir);
                }
            }

        } else {
            System.err.println("Elemento não foi adicionada\nNumero " + x + " ja existe na arvore");
        }
    }

    public void inserirPai(int x) {
        if (x > raiz.elemento) {
            inserirPai(x, raiz, raiz.dir);
        } else {
            inserirPai(x, raiz, raiz.esq);
        }
    }

    private void inserirPai(int x, No pai, No filho) {
        if (filho == null) {
            if (pai.elemento < x) {
                pai.dir = new No(x);
            } else {
                pai.esq = new No(x);
            }
        } else {
            if (noDiferente(filho.elemento, x)) {
                if (filho.elemento > x) {
                    inserirPai(x, filho, filho.esq);
                } else {
                    inserirPai(x, filho, filho.dir);
                }
            } else {
                System.err.println("Elemento não foi adicionada\nNumero " + x + " ja existe na arvore");

            }
        }
    }

    private boolean noDiferente(int x, int y) {
        return (x != y);
    }

    public boolean pesquisar(int x) {
        return pesquisar(x, raiz);
    }

    private boolean pesquisar(int x, No no) {
        boolean resp = false;
        if (no != null) {
            System.out.printf("%d ", no.elemento);
            if (no.elemento == x) {
                resp = true;
            } else {
                if (no.elemento < x) {
                    resp = pesquisar(x, no.dir);
                } else {
                    resp = pesquisar(x, no.esq);
                }
            }
        }
        return resp;
    }

    public void caminharPre() {
        if (raiz == null) {
         System.out.println("V");  
         return; 
        }
        caminharPre(raiz);
        System.out.println("");  

    }

    private void caminharPre(No no) {
        if (no != null) {
            System.out.printf("%d ", no.elemento);
            caminharPre(no.esq);
            caminharPre(no.dir);
        }
    }

    public void caminharPos() {
        if (raiz == null) {
         System.out.println("V");  
         return; 
        }
        caminharPos(raiz);
        System.out.println("");  

    }

    private void caminharPos(No no) {
        if (no != null) {
            caminharPos(no.esq);
            caminharPos(no.dir);
            System.out.printf("%d ", no.elemento);

        }
    }

    public void caminharCentral() {
        if (raiz == null) {
         System.out.println("V");  
         return; 
        }
        caminharCentral(raiz);
        System.out.println("");  
    
    }

    private void caminharCentral(No no) {
        if (no != null) {
            caminharCentral(no.esq);
            System.out.printf("%d ", no.elemento);
            caminharCentral(no.dir);
        }
    }
}

class No {

    int elemento;
    No esq;
    No dir;

    No(int x) {
        elemento = x;
        esq = null;
        dir = null;
    }
}

public class Main {

    public static void main(String[] args) {
        Arvore arvore = new Arvore();
        Scanner sc = new Scanner(System.in);

        for (String linha = sc.next(); !linha.equals("-1"); linha = sc.next()) {
            lerComando(sc, linha, arvore);
            if (!sc.hasNext()) {
                break;
            }
        }
        sc.close();
    }

    public static void lerComando(Scanner sc, String comando, Arvore arvore) {
        switch (comando) {
            case "I":

                inserir(sc.nextInt(), arvore);
                break;
            case "P":
                pesquisar(sc.nextInt(), arvore);
                break;
            case "PRE":
                caminharPre(arvore);
                break;
            case "POS":
                caminharPos(arvore);
                break;
            case "EM":
                caminharCentral(arvore);
                break;
            default:
                System.out.println("Comando não aceito");
        }
    }

    private static void inserir(int x, Arvore arvore) {
        arvore.inserir(x);
    }

    private static void pesquisar(int x, Arvore arvore) {
        System.out.println((arvore.pesquisar(x) ? "S" : "N"));
    }

    private static void caminharPre(Arvore arvore) {
        arvore.caminharPre();
    }

    private static void caminharPos(Arvore arvore) {
        arvore.caminharPos();
    }

    private static void caminharCentral(Arvore arvore) {
        arvore.caminharCentral();
    }

}

File: Periodo3/aeds2/estudo/atvLab/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct no_t
{
    int elemento;
    struct no_t *dir;
    struct no_t *esq;

} no;
typedef struct
{
    no *raiz;
} arvore;
int pesquisa(no *n, int x)
{
    if (n == NULL)
    {
        return 0;
    }
    printf("%d ", n->elemento);
    int resp = 0;
    if (n->elemento > x)
    {
        if (n->esq != NULL)
        {
            resp = pesquisa(n->esq, x);
        }
    }
    else
    {
        if (n->elemento < x)
        {
            if (n->dir != NULL)
            {
                resp = pesquisa(n->dir, x);
            }
        }
        else
        {
            resp = 1;
        }
    }
    return resp;
}
void caminhar_pre(no *n)
{
    if (n == NULL)
    {

        return;
    }

    printf("%d ", n->elemento);
    caminhar_pre(n->esq);
    caminhar_pre(n->dir);
}

void caminhar_pos(no *n)
{
    if (n == NULL)
    {

        return;
    }
    caminhar_pos(n->esq);
    caminhar_pos(n->dir);
    printf("%d ", n->elemento);
}

void caminhar_central(no *n)
{
    if (n == NULL)
    {

        return;
    }
    caminhar_central(n->esq);
    printf("%d ", n->elemento);
    caminhar_central(n->dir);
}

void inserir(arvore *a, no *n, int x)
{
    if (a->raiz == NULL)
    {
        n = (no *)malloc(sizeof(no));
        n->elemento = x;
        n->dir = NULL;
        n->esq = NULL;
        a->raiz = n;
    }
    else
    {

        if (n->elemento > x)
        {
            if (n->esq == NULL)
            {
                n->esq = (no *)malloc(sizeof(no));
                n->esq->dir = NULL;
                n->esq->esq = NULL;
                n->esq->elemento = x;
            }
            else
            {
                inserir(a, n->esq, x);
            }
        }
        else
        {
            if (n->dir == NULL)
            {
                n->dir = (no *)malloc(sizeof(no));
                n->dir->dir = NULL;
                n->dir->esq = NULL;
                n->dir->elemento = x;
            }
            else
            {
                inserir(a, n->dir, x);
            }
        }
    }
}

int main()
{
    arvore *a = (arvore *)malloc(sizeof(arvore));
    a->raiz = NULL;
    char c[4];
    int x;
    while (scanf("%s", c) != EOF)
    {
        if (strcmp(c, "I") == 0 || strcmp(c, "P") == 0)
        {
            scanf("%d", &x);
        }
        if (strcmp(c, "I") == 0)
        {
            inserir(a, a->raiz, x);
        }
        else
        {
            if (strcmp(c, "P") == 0)
            {
                printf("%c\n", ((pesquisa(a->raiz, x) == 1) ? 'S' : 'N'));
            }
            else
            {
                if (strcmp(c, "PRE") == 0)
                {
                    if (a->raiz == NULL)
                    {
                        printf("V");
                    }

                    caminhar_pre(a->raiz);
                    printf("\n");
                }
                else
                {
                    if (strcmp(c, "POS") == 0)
                    {
                        if (a->raiz == NULL)
                        {
                            printf("V");
                        }
                        caminhar_pos(a->raiz);
                        printf("\n");
                    }
                    else
                    {
                        if (a->raiz == NULL)
                        {
                            printf("V");
                        }
                        caminhar_central(a->raiz);
                        printf("\n");
                    }
                }
            }
        }
    }
}
